### ISSUE TITLE
METoken contract migration issue.

### ISSUE THREAD

**Issue by chainhead** [ethereumbook/ethereumbook#328]
I want to deploy MET to private network. I have created a separate project to deploy MET. The project is here: https://github.com/chainhead/MET that I will raise a PR for eventually.

I have edited my `truffle.js` to point to my private network. 

```
module.exports = {
	networks: {
		aws: { // Private network on AWS
			network_id: "15",
			host: "public-ip-address",
			port: 8545,
			gas: 6721975
		}
	}
};
```
The truffle compilation goes through. However, the migration does not give any information about successful deployment. Can you please suggest what may I be missing?

```
truffle migrate --network aws
Using network 'aws'.

Network up to date.

```

Also, I noted the following and fixing them in my repository.
- Truffle configuration file points to `Faucet.sol` in `tokens.asciidoc`. Shouldn't it be one in `METoken.sol`?
- Import error for `StandardToken.sol`.
- Require error for `artifacts.require("Migrations")`.
- Lint errors as per VSCode.

**Comment by fjrojasgarcia** [comment]
Only a very basic question, Have you created your [2_deploy_contracts.js](https://github.com/ethereumbook/ethereumbook/blob/develop/code/truffle/METoken/migrations/2_deploy_contracts.js) deployment file?

```
var METoken = artifacts.require("METoken");

module.exports = function(deployer) {
  // Deploy the METoken contract as our only task
  deployer.deploy(METoken);
};
```

**Comment by AyushyaChitransh** [comment]
Also, you would like to use another flag `--reset` if you have already deployed on that network and want to deploy again. 

    Network up to date.

This message is mostly there if there exists a `build` folder with the information of the network. And it has the information that the same contract has already been deployed on the network.

**Comment by chainhead** [comment]
@fjrojasgarcia 
Yes, I was missing the second deployment file. I added it and then upon migration, I get this issue.

```
Using network 'aws'.

/usr/local/lib/node_modules/truffle/build/cli.bundled.js:128857
var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;if(Module["ENVIRONMENT"]){if(Module["ENVIRONMENT"]==="WEB"){ENVIRONMENT_IS_WEB=true}else if(Module["ENVIRONMENT"]==="WORKER"){ENVIRONMENT_IS_WORKER=true}else if(Module["ENVIRONMENT"]==="NODE"){ENVIRONMENT_IS_NODE=true}else if(Module["ENVIRONMENT"]==="SHELL"){ENVIRONMENT_IS_SHELL=true}else{throw new Error("The provided Module['ENVIRONMENT'] value is not valid. It must be one of: WEB|WORKER|NODE|SHELL.")}}else{ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_IS_NODE=typeof process==="object"&&"function"==="function"&&!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_WORKER;ENVIRONMENT_IS_SHELL=!

Error: Expected parameter 'from' not passed to function.
    at /usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-expect/index.js:5:1
    at Array.forEach (<anonymous>)
    at Object.options (/usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-expect/index.js:3:1)
    at Object.run (/usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-migrate/index.js:122:1)
    at /usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-core/lib/commands/migrate.js:79:1
    at /usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-migrate/index.js:287:1
    at /usr/local/lib/node_modules/truffle/build/webpack:/~/truffle-migrate/index.js:115:1
    at done (/usr/local/lib/node_modules/truffle/build/webpack:/~/node-dir/lib/paths.js:27:1)
    at /usr/local/lib/node_modules/truffle/build/webpack:/~/node-dir/lib/paths.js:60:1
    at FSReqWrap.oncomplete (fs.js:167:5)
```

@AyushyaChitransh I will try the reset option once I have got past this error.

**Comment by fjrojasgarcia** [comment]
Only to ensure .. have you installed the Solidity dependences?

`npm install zeppelin-solidity`

**Comment by chainhead** [comment]
@fjrojasgarcia Yes.

**Comment by fjrojasgarcia** [comment]
I would also remove or temporary rename the build directory to avoid possible inconsistences.

**Comment by chainhead** [comment]
@fjrojasgarcia Removed the build folder, compiled and migrated; no luck. Following up on the truffle gitter channel - https://gitter.im/ConsenSys/truffle

**Comment by chainhead** [comment]
@fjrojasgarcia 
I was missing the `from` field, to hold account address, in `truffle.js`. I had to use an account address that was used for mining because only this account (ether base) can be expected to have funds to deploy contracts, etc. Also, I had to edit the `1_initial_migration.js` to apply password to unlock account. 

For now, I can see that the contract is getting deployed, but does not seem to complete.
````
truffle migrate --network aws 
Using network 'aws'.

Running migration: 1_initial_migration.js
  Deploying Migrations...
  ... 0x3d04e043f82e323fb326a75f6da382e6687be1df4d00f026207f1b8e4d892fe8
````
And, on the mining node:
````
INFO [03-26|23:15:55] Submitted contract creation              fullhash=0x3d04e043f82e323fb326a75f6da382e6687be1df4d00f026207f1b8e4d892fe8 contract=0xD33b0dCFFA52D2188E22BD01826d063265ec3e83
````

Questions:

1. The general flow seems to be like this. Create a new account on the mining node because, this is the only node, where mining rewards will be credited with ether that will in turn be used to fund contract deployments, etc. Correct?
2. Deploy `Faucet.sol` with this account address - otherwise, we get an 'insufficient gas' message. Later, invoke withdraw function to self-fund.
3. What is the address shown in console where `truffle migrate --network` is run? Contract location? (See above).
4. If yes, then should it not match with the address shown in the mining node output? (See above).
5. Is `fullhash` the hash of the contract ABI?
